chans/httpserver.go

```diff
@@ -0,0 +1,228 @@
+		Error   string              `json:"error,omitempty"`
+	}
+
+	type Response struct {
+		Headers       map[string][]string `json:"headers,omitempty"`
+		Body          interface{}         `json:"body,omitempty"`
+		StatusCode    int                 `json:"statuscode,omitempty"`
+		Serialization *dsl.Serialization  `json:"serialization,omitempty"`
+	}
+
+	punt := func(w http.ResponseWriter, err error) {
+		w.WriteHeader(501)
+		m := map[string]interface{}{
+			"error": err.Error(),
+		}
+		js, err := json.Marshal(&m)
+		if err != nil {
+			js = []byte(fmt.Sprintf(`{"error":"%s"}`, err.Error())) // Too optimistic
+		}
+
+		w.Write(js)
+	}
+
+	f := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		payload := &Payload{
+			Path:    r.URL.Path,
+			Headers: r.Header,
+			Method:  r.Method,
+		}
+
+		if r.Method == http.MethodPost {
+			bs, err := ioutil.ReadAll(r.Body)
+			if err != nil {
+				punt(w, err)
+				return
+			}
+
+			if 0 < len(bs) && c.opts.ParseJSON {
+				var body interface{}
+				if err := json.Unmarshal(bs, &body); err != nil {
+					punt(w, err)
+					return
+				}
+				payload.Body = body
+			} else {
+				payload.Body = string(bs)
+			}
+		}
+
+		js, err := json.Marshal(payload)
+		if err != nil {
+			punt(w, err)
+			return
+		}
+
+		req := dsl.Msg{
+			Topic:   r.URL.Path,
+			Payload: string(js),
+		}
+
+		select {
+		case <-ctx.Done():
+		case c.reqs <- req:
+			select {
+			case <-ctx.Done():
+			case resp := <-c.resps:
+				r := &Response{
+					StatusCode:    200, // ToDo: opt
+					Serialization: &dsl.DefaultSerialization,
+				}
+				if err := json.Unmarshal([]byte(resp.Payload), &r); err != nil {
+					w.WriteHeader(501)
+					w.Write([]byte(err.Error() + " on payload"))
+					return
+				}
+				body, err := r.Serialization.Serialize(r.Body)
+				if err != nil {
+					w.WriteHeader(501)
+					w.Write([]byte(err.Error() + " on response"))
+					return
+				}
+				w.WriteHeader(r.StatusCode)
+				// ToDo: Check err, bytes written.
+				w.Write([]byte(body))
+			}
+		}
+	})
+
+	c.server = &http.Server{
+		Addr:           addr,
+		Handler:        f,
+		ReadTimeout:    10 * time.Second, // ToDo: opt
+		WriteTimeout:   10 * time.Second, // ToDo: opt
+		MaxHeaderBytes: 1 << 16,          // ToDo: opt
+	}
+
+	go func() {
+		// ToDo: Report failure to listen better.
+		if err := c.server.ListenAndServe(); err != nil {
+			ctx.Logf("httpserver ListenAndServe error: %v", err)
+		}
+	}()
+
+	return nil
+}
+
+func (c *HTTPServer) Close(ctx *dsl.Ctx) error {
+	return c.server.Close()
+}
+
+func (c *HTTPServer) Sub(ctx *dsl.Ctx, topic string) error {
+	return dsl.Brokenf("%T doesn't support 'sub'", c)
+}
+
+func (c *HTTPServer) Pub(ctx *dsl.Ctx, m dsl.Msg) error {
+	ctx.Logf("%T Pub", c)
+	return c.To(ctx, m)
+}
+
+func (c *HTTPServer) Recv(ctx *dsl.Ctx) chan dsl.Msg {
+	return c.reqs
+}
+
+func (c *HTTPServer) Kill(ctx *dsl.Ctx) error {
+	return dsl.Brokenf("%T doesn't support 'Kill'", c)
+}
```

Comment:
I think supporting a `Kill` step for the httpserver channel could lead to some interesting chaos plax tests